### PR TITLE
[lang] add timeslice() operator

### PR DESCRIPTION
**Pull request by tstack**
This change is a first pass at adding the timeslice() operator.  It seems to do the job, most of the work is done by the chrono crate's duration_trunc() method.

Related to #10

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -364,6 +364,27 @@ In addition to columns, `sort` can also sort an arbitrary expressions.
 * | json | sort by length(endpoint_url)
 ```
 
+##### Timeslice
+`timeslice(<timestamp>) <duration> [as <field>]`: Truncates a timestamp to the
```
Comment (rcoh) [inline README.md:368]: It looks like duration is optional but defaults to 1h. I think it's probably better to just error during typecheck if a duration isn't set and remove the default.

File: src/typecheck.rs [hunk 2]
```diff
@@ -279,6 +279,15 @@ impl TypeCheck<Box<dyn operator::OperatorBuilder + Send + Sync>>
                     .map(|e| e.type_check(error_builder))
                     .transpose()?,
             ))),
+            lang::InlineOperator::Timeslice {
+                input_column,
+                duration,
+                output_column,
+            } => Ok(Box::new(operator::Timeslice::new(
+                input_column.type_check(error_builder)?,
+                duration.unwrap_or_else(|| chrono::Duration::hours(1)),
```
Comment (rcoh) [inline src/typecheck.rs:288]: probably better to error here—a default of 1h is probably not what people want?

Although if that's actually a sensible default we could just fix the docs.

File: src/operator.rs [hunk 3]
```diff
@@ -988,6 +995,50 @@ impl UnaryPreAggFunction for FieldExpressionDef {
...
+    }
+}
+
+impl UnaryPreAggFunction for Timeslice {
+    fn process(&self, rec: Record) -> Result<Option<Record>, EvalError> {
+        let inp: data::Value = self.input_column.eval(&rec.data)?;
+
+        match inp {
+            data::Value::DateTime(dt) => {
+                let rounded = dt.duration_trunc(self.duration).unwrap();
```
Comment (rcoh) [inline src/operator.rs:1025]: that's a really handy function!

**Review by rcoh — CHANGES_REQUESTED** [review]
mostly LGTM, just want to get the docs to match the engine on optional vs. non optional

File: README.md (hunk 1 above)
Comment (tstack) [reply README.md:368]: changed to raise an error

**Review by rcoh — APPROVED** [review]

**Pull request merged**